### ISSUE TITLE
Bug: Creating new physical network interface with a pre-defined name fails to rename the interface after apply

### ISSUE THREAD

**Issue by vishvikkrishnan** [maas/terraform-provider-maas#165]
The `terraform plan` lists that the name would be applied to the created physical network interface as follows:

```hcl
# maas_network_interface_physical.configure_nic["node01"] will be created
  + resource "maas_network_interface_physical" "configure_nic" {
      + id          = (known after apply)
      + mac_address = "8c:ae:4c:df:50:7c"
      + machine     = (known after apply)
      + mtu         = (known after apply)
      + name        = "usb0"
      + tags        = (known after apply)
    }
```

However, after the `terraform apply`, the state indicates shows the resource with the original interface name as follows:

```hcl
# maas_network_interface_physical.configure_nic["node01"]:
resource "maas_network_interface_physical" "configure_nic" {
    id          = "176"
    mac_address = "8c:ae:4c:df:50:7c"
    machine     = "73ywba"
    mtu         = 1500
    name        = "enx8cae4cdf507c"
    tags        = []
    vlan        = 89
}
```

Additionally, trying to reapply the same plan, does update the name through an in-place update operation.

```hcl
  # maas_network_interface_physical.configure_nic["node01"] will be updated in-place
  ~ resource "maas_network_interface_physical" "configure_nic" {
        id          = "176"
      ~ name        = "enx8cae4cdf507c" -> "usb0"
        tags        = []
        # (4 unchanged attributes hidden)
    }
```
The state is as expected after re-applying the same plan:

```hcl
# maas_network_interface_physical.configure_nic["node01"]:
resource "maas_network_interface_physical" "configure_nic" {
    id          = "176"
    mac_address = "8c:ae:4c:df:50:7c"
    machine     = "73ywba"
    mtu         = 1500
    name        = "usb0"
    tags        = []
    vlan        = 17
}
```

The name should ideally update during the first apply as well, and there should be no need to apply the same plan again (otherwise, there's no idempotency.)

**Comment by ednxzu** [comment]
Can confirm, I have been running into the same issue.

**Comment by skatsaounis** [comment]
Hi @vishvikkrishnan,

Thank you for providing valuable feedback for the provider. Let me try to investigate if this is an issue of the provider or something MAAS internal. Let me ask two questions for clarification:

- `enx8cae4cdf507c` do you have an idea where this name is declared or how it is produced?
- You are applying the terragrunt values of the linked PR?

In the meantime I will try to run the latest version of the provider on my dev MAAS from master branch to try to replicate what you are facing

**Comment by skatsaounis** [comment]
Ok that was quicker than anticipated. I replicated the issue. It appears only if you are trying to create an interface that already exists. There is a [check](https://github.com/maas/terraform-provider-maas/blob/master/maas/resource_maas_network_interface_physical.go#L107-L112) that is turning the create into no-op. The correct behavior would be to error and exit.

So what is actually happening is the following:

1. Terraform will try to create. It will find an existing interface with the same macaddress. It will turn the create into no-op.
2. Terraform will perform a read as a last step of the create function and will fill its state with the existing interface information from MAAS. In your case: `enx8cae4cdf507c`
3. On next apply, Terraform will detect a drift and will try to run the update function.
4. The update function is renaming `enx8cae4cdf507c` to `usb0`.

Correct behavior would be:

1. User tries to create an interface with a macaddress that already exists. Error out to the user.
2. User is importing the existing interface to the terraform resource.
3. User runs apply and the update function changes the existing interface fields.

**Comment by ednxzu** [comment]
Hello, This shouldn't error out imo, because it can be linked to the creation of a maas_machine object.


An example module we have to provision a complete machine.
```hcl
resource "maas_machine" "this" {
  hostname        = var.server_name
  domain          = var.server_domain_name
  pxe_mac_address = var.server_interfaces[var.maas_pxe_interface].interface_mac_addr

  architecture = var.maas_server_architecture
  pool         = var.maas_server_pool
  zone         = var.maas_server_zone

  power_type       = var.maas_power_type
  power_parameters = jsonencode(var.maas_power_options)
}

resource "maas_network_interface_physical" "this" {
  for_each = var.server_interfaces

  mac_address = each.value.interface_mac_addr
  machine     = maas_machine.this.id
  mtu         = each.value.vlan_mtu
  name        = each.key
  tags        = []
  vlan        = each.value.maas_vlan_id
}

resource "maas_network_interface_link" "this" {
  for_each = var.server_interfaces

  machine           = maas_machine.this.id
  network_interface = maas_network_interface_physical.this[each.key].id
  subnet            = each.value.maas_subnet_id
  mode              = "STATIC"
  ip_address        = each.value.ip_address
  default_gateway   = true
}
```

this is called like: 
```hcl
module "os_control_03" {
  source = "../../../../../modules/full_cloud_server"

  server_name              = "os-control-03"
  server_domain_name       = maas_dns_domain.redacted.name
  maas_server_pool         = maas_resource_pool.poweredge_7525.name
  maas_server_zone         = "eu-west-1"
  maas_server_architecture = "amd64/generic"
  maas_pxe_interface       = "public"

  server_interfaces = {
    public = {
      interface_mac_addr = "<redacted>"
      ip_address         = "10.1.100.43"
      maas_vlan_id       = maas_vlan.vlan_100.id
      maas_subnet_id     = maas_subnet.subnet_vlan_100.id
      vlan_mtu           = maas_vlan.vlan_100.mtu
    }
    private = {
      interface_mac_addr = "<redacted>"
      ip_address         = "10.1.110.43"
      maas_vlan_id       = maas_vlan.vlan_110.id
      maas_subnet_id     = maas_subnet.subnet_vlan_110.id
      vlan_mtu           = maas_vlan.vlan_110.mtu
    }
  }
}
```

I'm not sure of the course of action, but erroring out in case the interface already exist would be a pretty bad regression.

In the above example, machine gets created, booted up and provisioned, then the interfaces are created (rather edited, since they have been detected during provisioning).

I think (? maybe I'm missing something) the interfaces should be edited regardless ? Having to import interfaces into the state makes it so that it's pretty much impossible to streamline the provisioning to a CI run or smthg like that, since you would have to have manual steps in the middle. I'm curious about what sort of undesired side effect could happen if the modify operation happens regardless.

**Comment by skatsaounis** [comment]
Hi @ednxzu. Thank you for guiding my thoughts here. After spinning around it for a while I created the linked PR which hopefully makes the provider behave the expected Terraform way. Please give us some time to review it. In the meantime, both you and @vishvikkrishnan could try to to build the provider from this branch and let me know if it is working for you :slightly_smiling_face:

**Comment by skatsaounis** [comment]
This is a snippet that you can use:

```tf
resource "maas_network_interface_physical" "test_physical" {
  machine     = maas_machine.my_machine.id
  mac_address = tolist(maas_machine.my_machine.network_interfaces)[0].mac_address
  name        = "updated0"
}
```

**Comment by hamatuni** [comment]
@skatsaounis, I am experiencing a similar issue on my side. After building the terraform provider from your branch, I can confirm that the issue is resolved and everything works as expected.
Thank you for your effort.

**Comment by vishvikkrishnan** [comment]
Hi @skatsaounis, thanks for the PR to fix this issue! I built the provider locally and confirmed the fix works by importing the existing NICs into the TF state and updating the interface name as well. But then, it looks like setting or updating the `vlan` field of the imported physical interface is not possible anymore. After a successful apply, the state still maintains the imported value for the `vlan` field, making subsequent `terraform apply` operations suggest an recurring in-place-update operation that is never applied.

In the below example, I set the `vlan` for the first interface while, left it empty for the second interface.
```hcl
Terraform will perform the following actions:

  # maas_network_interface_physical.configure_nic["node01ob48-enp0s25"] will be created
  + resource "maas_network_interface_physical" "configure_nic" {
      + id          = (known after apply)
      + mac_address = "b8:ae:ed:73:38:40"
      + machine     = "node01ob48"
      + mtu         = 1500
      + name        = "enp0s25"
      + tags        = (known after apply)
      + vlan        = 0
    }

  # maas_network_interface_physical.configure_nic["node01ob48-usb0"] will be created
  + resource "maas_network_interface_physical" "configure_nic" {
      + id          = (known after apply)
      + mac_address = "8c:ae:4c:df:50:7c"
      + machine     = "node01ob48"
      + mtu         = 1500
      + name        = "usb0"
      + tags        = (known after apply)
      + vlan        = (known after apply)
    }
```

When I run `terraform apply` again after a successful apply, I see the following output:
```hcl
maas_network_interface_physical.configure_nic["node01ob48-usb0"]: Refreshing state... [id=249]
maas_network_interface_physical.configure_nic["node01ob48-enp0s25"]: Refreshing state... [id=241]

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # maas_network_interface_physical.configure_nic["node01ob48-enp0s25"] will be updated in-place
  ~ resource "maas_network_interface_physical" "configure_nic" {
        id          = "241"
        name        = "enp0s25"
        tags        = []
      ~ vlan        = 5 -> 0
        # (3 unchanged attributes hidden)
    }
```
The above plan seems to never take effect on the interface, however.
Is there any way to update the `vlan` field of the imported interface?

**Comment by skatsaounis** [comment]
Hi @vishvikkrishnan , The situation you are describing is not a bug in terms of updating the vlan.

First of all, you need to set the database ID of the VLAN to that field as it is described in the [docs](https://registry.terraform.io/providers/maas/maas/latest/docs/resources/network_interface_physical#vlan).

So to properly set the vlan you can do something like the following, given that you are not aware of the database ID:

```tf
data "maas_fabric" "tf_test_fabric" {
  name = "fabric-1"
}

data "maas_vlan" "tf_test_vlan" {
  fabric = data.maas_fabric.tf_test_fabric.id
  vlan   = 0
}

resource "maas_network_interface_physical" "test_physical" {
  machine     = "sacred-horse"
  mac_address = "00:16:3e:01:44:5f"
  vlan        = data.maas_vlan.tf_test_vlan.id
  name        = "usb0"
}
```

And the plan is saying the proper thing (not vid `0` but database ID `2`):

```bash
data.maas_fabric.tf_test_fabric: Reading...
data.maas_fabric.tf_test_fabric: Read complete after 0s [id=1]
data.maas_vlan.tf_test_vlan: Reading...
data.maas_vlan.tf_test_vlan: Read complete after 0s [id=2]
maas_network_interface_physical.test_physical: Refreshing state... [id=11]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # maas_network_interface_physical.test_physical will be updated in-place
  ~ resource "maas_network_interface_physical" "test_physical" {
        id          = "11"
        name        = "usb0"
        tags        = []
      ~ vlan        = 3 -> 2
        # (3 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

Secondly, let's try to understand why you didn't receive an error when you used `0`. This is because of the `gomaasclient` implementation:

```go
// NetworkInterfaceUpdateParams is the parameters for the NetworkInterfaces update_bond POST operation.
type NetworkInterfaceUpdateParams struct {
...
	VLAN               int    `url:"vlan,omitempty"`
...
}
```

That `omitempty` will remove the choice of `0` when it will try to create the update params. So MAAS will never get the `0`. Otherwise, it would return 400:

```bash
ServerError: 400 Bad Request ({"vlan": ["Select a valid choice. That choice is not one of the available choices."]})
```

**Comment by skatsaounis** [comment]
tldr; please use the snippet I shared with you for properly setting the VLAN Database ID and we will think what we should or shouldn't do for the gomaasclient issue. :beer:

**Comment by vishvikkrishnan** [comment]
Crystal clear! Thanks a lot for the detailed explanation, @skatsaounis! All good then :)
